### PR TITLE
feat(tetris/core): pure logic layer + unit tests

### LINKED ISSUE #3: Tetris core: pure logic layer + unit tests
### Problem
Game rules tangled with rendering are untestable, hard to reason about, and impossible to reuse. Tetris in particular has many subtle rules (SRS rotation, wall kicks, T-spin recognition, scoring tiers) that must be airtight before any rendering code touches them.

### Goal
Implement Tetris's complete rule layer as **pure GDScript**, zero `Node`/`Engine` references. Drop-in consumable by the rendering layer in #4.

### Scope
- 10×20 visible playfield + 20-row buffer above (per Tetris Guideline).
- 7 standard tetrominoes (I, O, T, S, Z, J, L).
- SRS (Super Rotation System) — 4 rotation states per piece, full wall-kick tables for JLSTZ and the larger I-piece kicks.
- 7-bag random generator, seedable.
- Lock delay (with reset-on-move, capped at N moves per Tetris Guideline).
- Line clear detection (singles → tetris).
- Scoring (single, double, triple, tetris, T-spin {mini, normal} {empty, single, double, triple}, back-to-back, combo).
- Hold piece.
- Soft drop, hard drop.
- Top-out / lock-out detection.

### Non-goals
- Rendering, animation, HUD (→ #4).
- Sound, haptics, settings persistence (→ #5).
- AI, leaderboards, multiplayer.

### Acceptance
1. Every SRS rotation transition for every piece is unit-tested (4 states × 7 pieces × 2 directions = 56 minimum cases, but I/O/Z/S have rotational symmetry so unique outcomes are fewer; tests assert *resulting* cells, not just direction labels).
2. Every wall-kick offset in the SRS table is exercised at least once with a constructed scenario.
3. T-spin detection covers: corners-3 rule, mini vs full, kick-induced T-spins, T-spin while hold-swap.
4. Scoring formula matches the [Tetris Wiki Guideline](https://tetris.wiki/Scoring) tables exactly; back-to-back chains and combos verified.
5. 7-bag fairness: over 70 pulls (10 bags), every piece appears exactly 10 times. Seed determinism: same seed → same sequence.
6. `Board` exposes deterministic state-snapshot for replay/integration tests (next PR).
7. CI green; coverage of `scripts/tetris/core/` is effectively 100% by tests.

### Files
```
godot/scripts/tetris/core/
  piece_kind.gd           # enum + spawn shapes + rotation states
  piece.gd                # value object: kind, rotation, origin (col,row)
  bag.gd                  # 7-bag, seedable
  kicks.gd                # SRS kick tables (JLSTZ + I)
  board.gd               # 10x40 grid, lock-in, line-clear, T-spin detection
  scoring.gd              # scoring formula + B2B + combo state
  game_state.gd           # composes the above; pure controller, no _process
godot/tests/unit/tetris/
  test_piece_rotations.gd
  test_kicks.gd
  test_bag.gd
  test_board_line_clear.gd
  test_board_t_spin.gd
  test_scoring.gd
  test_game_state.gd
```

### Public API (`GameState`)
```gdscript
class_name TetrisGameState extends RefCounted

signal piece_locked(rows_cleared: int, t_spin: int)  # t_spin: 0|1=mini|2=full
signal game_over

static func create(seed: int) -> TetrisGameState

func tick(now_ms: int) -> void                       # gravity + lock-delay timing
func move(dx: int) -> bool
func rotate(dir: int) -> bool                        # +1 cw, -1 ccw
func soft_drop() -> void
func hard_drop() -> int                              # returns rows fallen
func hold() -> bool

func snapshot() -> Dictionary                        # for tests/replays
func current_piece() -> Piece
func ghost_position() -> Vector2i
func score() -> int
func level() -> int
func lines_cleared() -> int
func is_game_over() -> bool
```

### Data shapes
- `PieceKind` is an enum-ish constant block on `piece_kind.gd` plus a `SHAPES` dict mapping (kind, rotation) → `Array[Vector2i]` of cell offsets.
- `Board.cells: PackedInt32Array` of length 10*40, 0 = empty, 1..7 = piece kind. `idx(col,row) -> col + row*10`.
- Kick tables: `Kicks.JLSTZ` and `Kicks.I` — `Dictionary[String, Array[Vector2i]]` keyed `"0->R"`, `"R->0"`, … (5 kicks each per Guideline).

### Determinism + testability
- All randomness through `Bag` which holds a seeded `RandomNumberGenerator`.
- `tick` takes `now_ms` parameter — never reads engine time. Tests inject monotonic timestamps.
- No globals, no singletons, no autoloads referenced.

### Tests — coverage map
- **Rotations**: enumerate all 56 piece-state pairs, assert resulting offsets.
- **Kicks**: for each piece (JLSTZ uses one table, I uses another, O has no kicks), construct a board state where vanilla rotation collides; assert each of the 5 kick attempts is tried in order and the correct one is chosen.
- **Bag**: 70 pulls = 10 of each; same seed → same sequence; reseed mid-stream resumes correctly.
- **Line clear**: 1, 2, 3, 4 simultaneous clears; clear with empty row above; T-spin clears.
- **T-spin**: corners-3 rule (3 of the 4 corners around T center occupied); mini vs full distinction; kick-induced T-spin (only the (-1,-2) kick promotes mini → full per Guideline).
- **Scoring**: every row in the [Guideline scoring table](https://tetris.wiki/Scoring); B2B chain of 3; combo ladder up to 5.
- **Game over**: lock-out (piece can't spawn) and top-out (lock above the visible field).
- **Snapshot determinism**: same seed + same input sequence → identical snapshot.

### Risks
- **Kick table errors** — the I-piece kicks are non-obvious and easy to mistype. Mitigation: copy the table from the [Tetris Wiki SRS page](https://tetris.wiki/Super_Rotation_System), constants-only, no derivations; one test per row of the table.
- **T-spin classification edge cases** — particularly *kick-induced* T-spins. Mitigation: dedicated test fixtures with hand-drawn boards (commented ASCII art in tests).
- **Floating-point drift in gravity** — avoid floats; gravity is a frames-per-row integer that decreases with level.

### Commit sequence
1. `feat(tetris/core): piece kinds + SRS rotation states + tests`
2. `feat(tetris/core): SRS kick tables + tests`
3. `feat(tetris/core): board + line clear + tests`
4. `feat(tetris/core): T-spin detection + tests`
5. `feat(tetris/core): 7-bag + tests`
6. `feat(tetris/core): scoring (B2B + combo) + tests`
7. `feat(tetris/core): GameState composition + lock delay + tests`

---

## Sub-agent review — refinements folded in

Reviewed by `Plan` sub-agent. Resolutions:

### Tetris rules — explicit decisions
- **Lock delay**: 500 ms initial; movement/rotation that succeeds *resets the timer*, capped at **15 successful resets per piece**. Any successful downward step resets the cap counter — Guideline-compliant infinity prevention.
- **Line-clear gravity**: **naive** (per-row) — no sticky/cascade variant.
- **Spawn**: rows 21–22 (above visible field). I-piece spawns horizontal in row 21; J/L/S/T/Z spawn flat side down in row 21 spanning cols 4–6; O spawns in cols 5–6.
- **Game-over reasons** (separate enum): `BLOCK_OUT` (spawn collision), `LOCK_OUT` (piece locks entirely above row 20), `TOP_OUT` (line-clear lift). `game_over(reason)` signal.
- **Soft drop scoring**: 1 point per cell. **Hard drop**: 2 points per cell. Soft drop accumulates per-cell as gravity advances while the soft-drop flag is set.
- **Hold rules**: one use per piece (until lock); first hold of a game pulls the next from the queue.
- **T-spin classification**: front/back-corner rule. Two of the three "front" corners (relative to T-pointing direction) plus one "back" corner occupied = full T-spin. One front + two back = mini. Kick #5 (the (-1,-2) for JLSTZ, equivalent for I) override: any successful T rotation that used kick #5 = full T-spin regardless of corners.
- **Bag fairness reframed**: "every 7 consecutive pulls form a permutation of {I,O,T,S,Z,J,L}" (not "70 = 10×7", which is only true at bag boundaries).

### API revisions
- **`LockResult` returned via `piece_locked`** signal, not loose params:
  ```gdscript
  signal piece_locked(result: Dictionary)
  # result keys: kind:int, rows:int, t_spin:int (0|1=mini|2=full),
  #              b2b:bool, combo:int, perfect_clear:bool
  ```
- **`move/rotate` return small dict** `{success: bool, kick_index: int}` (kick_index = -1 if no kick used or move was translation).
- **`soft_drop()` returns int** = cells dropped this call. Continuous soft-drop is the caller's responsibility to accumulate scoring per-tick.
- **Add `next_queue(n: int = 5) -> Array[int]`** for Next renderer.
- **Add `held_piece() -> int` and `can_hold() -> bool`**.

### Layering — DAS/ARR location
- DAS/ARR lives in **InputManager (#2)**, not core. Core's `move(dx)` is one move per call; the input layer paces the calls.

### Gravity table location
- New file `godot/scripts/tetris/core/levels.gd` with the Guideline level→frames-per-row table. Used by `game_state.gd`.

### Coverage gate
- Drop "effectively 100%". Replace with: "every public method of every core file has at least one direct test; CI fails if any test in `tests/unit/tetris/` is skipped."

### Tests added/clarified
- Soft-drop continuous: 5 ticks of soft drop = 5 cells = 5 score points.
- Block-out vs lock-out: dedicated fixtures.
- T-spin via kick #5 override: dedicated test.
- Hold one-use-per-piece: rejected second-hold attempt.
- Snapshot-determinism: same seed + same input list → identical snapshot (covers reproducibility for #7's integration test).

### Open questions resolved
- `tick(now_ms)` advances at **variable** `now_ms` deltas; gravity uses an integer accumulator (`gravity_ms_owed += dt`), no floats anywhere.
- Next-queue length exposed: 5.
- Level/gravity curve: Tetris Worlds (since it's the most-cited Guideline curve).

### PR DESCRIPTION
**Pull request by lubobill1990**
Closes #3

Pure GDScript Tetris rule engine — no Node, no Engine, no autoloads — composed from per-concern modules with one test file each.

## What's in the layer

- **piece_kind.gd** + **piece.gd** — SRS shapes & rotation states for all 7 tetrominoes; piece value object.
- **kicks.gd** — SRS wall-kick tables (JLSTZ + I) verbatim from the Guideline.
- **board.gd** — 10×40 grid (20 visible + 20 buffer), placement queries, lock-in, naive line-clear gravity, snapshot/restore.
- **t_spin.gd** — 3-corner T-spin classifier (front/back rule + kick-5 → full override).
- **bag.gd** — seedable 7-bag, peek for Next preview, mid-stream reseed.
- **scoring.gd** — Guideline base table (single/double/triple/tetris, T-spin, T-spin mini), B2B 1.5x for difficult clears, 50×combo×level combo bonus, perfect-clear bonus, soft/hard-drop accumulators.
- **levels.gd** — pre-computed Tetris-Worlds gravity table (ms/row, integer math only).
- **game_state.gd** — composes everything; tick-driven gravity + 500 ms lock delay (15-reset cap, downward steps reset the cap); BLOCK_OUT/LOCK_OUT/TOP_OUT reasons; piece_locked / game_over signals.

## How tested
- All seven layers have a dedicated test file under `godot/tests/unit/tetris/`.
- Full suite: **130/130 passing locally** (1593 asserts) via the same headless command CI runs.
- Coverage of the dev plan's test map: rotations (per-state cell layouts), kicks (every JLSTZ + I transition + cross-kind uniformity), bag (per-bag-7 permutation, no-3-in-a-row, seed determinism), board (line-clear cases incl. tetris and split-clears + snapshot round-trip), T-spin (front/back rule for all 4 rotations, kick-5 override, walls/floor count as corners), scoring (every Guideline row, B2B chain of 3, combo ladder, drop scoring, level curve, perfect clears), game_state (move/rotate/hold/hard-drop/soft-drop, gravity tick at L1, lock-delay timing, snapshot determinism with same seed + same inputs, BLOCK_OUT and LOCK_OUT signal emission).

## Estimated effective diff
905 lines (total 1957, excluded 1052 — tests, .uid sidecar files).

⚠️ Exceeds 400 effective lines. Issue must carry the `large-pr-ok` label for auto-merge; otherwise human review will be required. The 905 lines are the unavoidable shape of "the entire Tetris rule layer" delivered as one unit per the dev plan, but a reviewer who'd rather see it split into 2-3 PRs (piece+kicks / board+t-spin / bag+scoring+game_state) can ask for that and I'll redo it.

### CODE REVIEW

File: godot/scripts/tetris/core/game_state.gd [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+	}
+
+# --- Internals ---
+
+func _advance_gravity(dt_ms: int) -> void:
+	if dt_ms <= 0:
+		return
+	var step_ms: int = Levels.ms_per_row(scoring.level)
+	if _soft_drop_active:
+		step_ms = max(1, step_ms / 20)  # 20G-equivalent ratio, but caller can also call soft_drop() directly
```
Comment (weavejamtom) [inline godot/scripts/tetris/core/game_state.gd]: `_soft_drop_active` is never assigned anywhere in the file (only read here, line 201). The whole `if _soft_drop_active` branch and the comment about "caller can also call soft_drop() directly" are dead — actual soft drop goes through `soft_drop()` (line 113-122). Either wire a setter and document the two paths, or delete the var + branch. Right now it's a footgun for the rendering layer (#4) which might assume setting `_soft_drop_active` does something.

File: godot/scripts/tetris/core/bag.gd [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+## Each pull returns a piece kind from a freshly shuffled permutation of
+## {I,O,T,S,Z,J,L}. After 7 pulls a new permutation is generated. Same seed
+## always produces the same sequence. Use `peek(n)` for the Next preview.
+
+const PieceKind := preload("res://scripts/tetris/core/piece_kind.gd")
+
+var _rng: RandomNumberGenerator
+var _bag: Array  # holds upcoming kinds in pull order; refilled when empty
+
+static func create(seed: int) -> TetrisBag:
```
Comment (weavejamtom) [inline godot/scripts/tetris/core/bag.gd]: `seed` shadows the global `seed()` builtin (Godot warns on this). Rename to `bag_seed` for parity with `game_state.gd`'s `game_seed`. Also propagate to `_init_seed` and `reseed`.

File: godot/scripts/tetris/core/scoring.gd [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+var combo: int = -1            # -1 = no combo active; first clear sets to 0
+var b2b: int = -1              # -1 = no B2B; first difficult clear sets to 0
+var perfect_clear_count: int = 0
+
+# Base points per (t_spin, rows_cleared). Values are pre-level-multiplied.
+const _BASE: Dictionary = {
+	# t_spin = NONE
+	0: { 0: 0, 1: 100, 2: 300, 3: 500, 4: 800 },
+	# t_spin = MINI: 100, 200, 400 for 0,1,2 lines (3+ rows is unreachable for mini).
+	1: { 0: 100, 1: 200, 2: 400, 3: 400 },
```
Comment (weavejamtom) [inline godot/scripts/tetris/core/scoring.gd]: Comment on the previous line says "3+ rows is unreachable for mini" but the dict has `3: 400`. Either the entry is unreachable and should be dropped (so a buggy caller fails loudly via the `get(rows, 0)` default), or the comment is wrong. Per Guideline a T-spin clearing 3 lines is always FULL, so I'd remove the `3: 400`.

File: godot/scripts/tetris/core/game_state.gd [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+	# field (row < VISIBLE_TOP), end the game with LOCK_OUT.
+	var all_above_visible: bool = true
+	for c in piece.cells():
+		if c.y >= Board.VISIBLE_TOP:
+			all_above_visible = false
+			break
+	if all_above_visible:
+		var locked_kind: int = piece.kind
+		piece = null
+		_end_game(REASON_LOCK_OUT)
```
Comment (weavejamtom) [inline godot/scripts/tetris/core/game_state.gd]: `_end_game` (which emits `game_over`) runs before `_emit_lock` (which emits `piece_locked`). HUD code that disconnects on `game_over` will miss the final lock signal — flip the order so `piece_locked` fires first, then `game_over`.

File: godot/scripts/tetris/core/game_state.gd [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+
+var board: Board
+var bag: Bag
+var scoring: Scoring
+
+var piece: Piece               # null if game over before respawn
+var held_kind: int = -1        # -1 = nothing held
+var hold_used: bool = false
+var _game_over: bool = false
+var _last_kick_index: int = -1 # records the kick used by the most recent rotate
```
Comment (weavejamtom) [inline godot/scripts/tetris/core/game_state.gd]: Worth noting that this is consumed only by T-spin classification at lock time (lines 264-265) — otherwise a future reader might think it's part of the move-result API.

**Review by weavejamtom — CHANGES_REQUESTED** [review]
Solid layered implementation; tests track the dev plan well and 130/130 green confirms the rule mechanics. Five findings — one is a real dead-branch bug, the rest are polish.

Note: per skill §4b, this PR can't auto-merge regardless of findings — the author is `lubobill1990`, not `weavejamtom`. Human merge is required after fixes.

### Findings
- **major** `game_state.gd:201-202` dead `_soft_drop_active` branch in `_advance_gravity`.
- **minor** `bag.gd:14` `seed` param shadows global `seed()`.
- **minor** `scoring.gd:26` MINI table entry `3: 400` contradicts the "3+ rows is unreachable for mini" comment above it.
- **minor** `game_state.gd:279` signal order on LOCK_OUT: `game_over` fires before `piece_locked`.
- **nit** `game_state.gd:40` doc could mention the kick index is consumed only by T-spin classification.

**Review by weavejamtom — APPROVED** [review]
Approved by gh-pr-review skill (human override of §4b: PR authored by lubobill1990 herself, reviewed as weavejamtom; effective diff 898 lines but content is naturally atomic — full Tetris core layer + matching unit tests — and prior 5 review findings are all addressed in 3ce7c7d). CI green.
